subsys/input/input_longpress.c

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright 2023 Google LLC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_input_longpress
+
+#include <zephyr/device.h>
+#include <zephyr/input/input.h>
+#include <zephyr/kernel.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(input_longpress, CONFIG_INPUT_LOG_LEVEL);
+
+struct longpress_config {
+	const struct device *input_dev;
+	const uint16_t *input_codes;
+	const uint16_t *short_codes;
+	const uint16_t *long_codes;
+	int num_codes;
+	int long_delays_ms;
+};
+
+struct longpress_data_entry {
+	const struct device *dev;
+	struct k_work_delayable work;
+	bool long_fired;
+};
+
+struct longpress_data {
+	struct longpress_data_entry *entries;
+};
+
+static void longpress_deferred(struct k_work *work)
+{
+	struct longpress_data_entry *entry = CONTAINER_OF(
+			work, struct longpress_data_entry, work);
+	const struct device *dev = entry->dev;
+	const struct longpress_config *cfg = dev->config;
+	struct longpress_data *data = dev->data;
+	int i;
+	uint16_t code;
+
+	for (i = 0; i < cfg->num_codes; i++) {
+		if (&data->entries[i] == entry) {
+			break;
+		}
+	}
+	if (i == cfg->num_codes) {
+		LOG_ERR("entry error %p", entry);
+		return;
+	}
+
+	code = cfg->long_codes[i];
+
+	input_report_key(dev, code, 1, true, K_FOREVER);
+
+	entry->long_fired = true;
+}
+
+static void longpress_cb(const struct device *dev, struct input_event *evt)
+{
+	const struct longpress_config *cfg = dev->config;
+	struct longpress_data *data = dev->data;
+	struct longpress_data_entry *entry;
+	int i;
+
+	if (evt->type != INPUT_EV_KEY) {
+		return;
+	}
+
+	for (i = 0; i < cfg->num_codes; i++) {
+		if (evt->code == cfg->input_codes[i]) {
+			break;
+		}
+	}
+	if (i == cfg->num_codes) {
+		LOG_DBG("ignored code %d", evt->code);
+		return;
+	}
+
+	entry = &data->entries[i];
+
+	if (evt->value) {
+		entry->long_fired = false;
+		k_work_schedule(&entry->work, K_MSEC(cfg->long_delays_ms));
+	} else {
+		k_work_cancel_delayable(&entry->work);
+		if (entry->long_fired) {
+			input_report_key(dev, cfg->long_codes[i], 0, true, K_FOREVER);
+		} else {
+			input_report_key(dev, cfg->short_codes[i], 1, true, K_FOREVER);
+			input_report_key(dev, cfg->short_codes[i], 0, true, K_FOREVER);
+		}
+	}
+}
+
+static int longpress_init(const struct device *dev)
+{
+	const struct longpress_config *cfg = dev->config;
+	struct longpress_data *data = dev->data;
+
+	if (cfg->input_dev && !device_is_ready(cfg->input_dev)) {
+		LOG_ERR("input device not ready");
+		return -ENODEV;
+	}
+
+	for (int i = 0; i < cfg->num_codes; i++) {
+		struct longpress_data_entry *entry = &data->entries[i];
+
+		entry->dev = dev;
+		k_work_init_delayable(&entry->work, longpress_deferred);
+	}
+
+	return 0;
+}
+
+#define INPUT_LONGPRESS_INIT(index)                                                                \
```

Comment:
s/index/inst, s/_INIT/_DEFINE

Reply:
Done, maybe one day we'll have a zephyr linter for this :-)